parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetReader.java

```diff
@@ -133,6 +148,83 @@ public T read() throws IOException {
     }
   }
 
+  private boolean readVectors(ColumnVector[] vectors, MessageType[] columns) throws IOException {
+    if (reader != null && reader.nextBatch(vectors, columns)) {
+      return true;
+    } else {
+      initReader();
+      if (reader == null) {
+        return false;
+      } else {
+        return readVectors(vectors, columns);
+      }
+    }
+  }
+
+  private boolean readVector(ObjectColumnVector<T> vector) throws IOException {
+    if (reader != null && reader.nextBatch(vector)) {
+      return true;
+    } else {
+      initReader();
+      if (reader == null) {
+        return false;
+      } else {
+        return readVector(vector);
+      }
+    }
+  }
+
+  /**
+   * Reads the next batch of rows. This method is used for reading complex types
+   * or arbitrary objects and calls the converters eventually to materialize the record.
+   * @param previous a row batch object to be reused by the reader if possible
+   * @param clazz the class of the record type that will be filled into the column vector
+   * @return the row batch that was read
+   * @throws java.io.IOException
+   */
+  public RowBatch nextBatch(RowBatch previous, Class<T> clazz) throws IOException {
+    RowBatch rowBatch = previous;
+    if (rowBatch == null) {
+      rowBatch = new RowBatch(new ColumnVector[] { ColumnVector.ofType(clazz) } );
+    }
+
+    return readVector((ObjectColumnVector<T>) rowBatch.getColumns()[0]) ? rowBatch : null;
+  }
+
+  /**
+   * Reads the next batch of rows. This method is used for reading primitive types
+   * and does not call the converters at all.
+   * @param previous a row batch object to be reused by the reader if possible
+   * @return the row batch that was read
+   * @throws java.io.IOException
+   */
+   public RowBatch nextBatch(RowBatch previous) throws IOException {
+     MessageType requestedSchema = readContext.getRequestedSchema();
+     List<ColumnDescriptor> columns = requestedSchema.getColumns();
+     int nColumns = columns.size();
+     ColumnVector[] columnVectors;
+
+     RowBatch rowBatch = previous;
+     if (rowBatch == null) {
+       rowBatch = new RowBatch(new ColumnVector[nColumns]);
+     }
+
+     columnVectors = rowBatch.getColumns();
+     MessageType[] columnSchemas = new MessageType[nColumns];
+     for (int i = 0; i < nColumns; i++) {
+       ColumnVector columnVector = columnVectors[i];
+       columnSchemas[i] = new MessageType(requestedSchema.getFieldName(i), requestedSchema.getType(i));
+
+       if (columnVector == null) {
+         columnVector = ColumnVector.from(columns.get(i));
+       }
+
+       rowBatch.getColumns()[i] = columnVector;
+     }
```

Comment:
I am not sure we should do it (at least in this PR) for two reasons. 1) We don't really know whether this is an overhead that we should worry about before measuring it. 2) It will complicate the logic here.

If you really think we should optimize this maybe we should open a jira for it and fix it once this gets merged.